File: python/golomb/golomb.py
import math

class Golomb:
    def __init__(self, m=2):
        assert m > 0

        self.m = m
        self.base2 = True if math.log2(m).is_integer() else False

    def set_m(self, m):
        assert m > 0

        self.m = m
        self.base2 = True if math.log2(m).is_integer() else False

    def encode(self, n):
        return self.base2encoder(n) if self.base2 else self.truncated_encoder(n)

    def base2encoder(self, n):
        
        negative = [0]
        if n < 0:
            n = abs(n)
            negative = [1]

        q = self.quocient(n, self.m)
        r = self.remainder(n, self.m)

        unary_code = self.unary_code(q)
        binary_code = self.decimal_to_binary(r, 2)

        golomb_code = unary_code + binary_code
        return negative + golomb_code
    
    def truncated_encoder(self, n):
        negative = [0]
        if n < 0:
            n = abs(n)
            negative = [1]

        b = math.ceil(math.log2(self.m))

        q = self.quocient(n, self.m)
        r = self.remainder(n, self.m)
        
        unary_code = self.unary_code(q)

        first_values = 2**b - self.m
        if r < first_values:
            binary_code = self.decimal_to_binary(r, b - 1)
        else:
            binary_code = self.decimal_to_binary(r + 2**b - self.m, b)
        
        golomb_code = unary_code + binary_code

        return negative + golomb_code
    
    def stream_decoder(self, bitstream):
        decoded = []
        while True:
            sign = bitstream[0]
            bitstream = bitstream[1:]

            if bitstream[0] == 0:
                unary_code = [bitstream[0]]
                binary_code = bitstream[1:(1 + math.ceil(math.sqrt(self.m)))]
                decimal = self.decode([sign] + unary_code + binary_code)
                decoded.append(decimal)
                bitstream = bitstream[(1 + math.ceil(math.sqrt(self.m))):]
                if not bitstream:
                    break
            else:
                unary_code = []
                while True:
                    bit = bitstream.pop(0)
                    unary_code.append(bit)
                    if bit == 0:
                        break
                binary_code = bitstream[0:math.ceil(math.sqrt(self.m))]
                decimal = self.decode([sign] + unary_code + binary_code)
                decoded.append(decimal)
                bitstream = bitstream[math.ceil(math.sqrt(self.m)):]
                if not bitstream:
                    break

        return decoded

    
    def decode(self, bitstream):
        assert len(bitstream) > 0
        return self.base2decoder(bitstream) if self.base2 else self.truncated_decoder(bitstream)

    def isNegative(self, bitstream):
        return bitstream[0] == 1

    def base2decoder(self, bitstream):
        
        negative = self.isNegative(bitstream)
        bitstream = bitstream[1:]

        if bitstream[0] == 0:
            q = 0
            r = self.binary_to_decimal(bitstream[1:])
        else:
            i = 0
            while bitstream[i] == 1:
                i += 1
            
            unary_code = bitstream[0:i+1]
            binary_code = bitstream[i+1:]
            
            q = i
            r = self.binary_to_decimal(binary_code)

        return r + q * self.m if not negative else -1 * (r + q * self.m)
    
    def truncated_decoder(self, bitstream):
        negative = self.isNegative(bitstream)
        bitstream = bitstream[1:]

        b = math.ceil(math.log2(self.m))

        if bitstream[0] == 0:
            q = 0
            binary_code = bitstream[1:]
        else:
            i = 0
            while bitstream[i] == 1:
                i += 1

            q = i
            binary_code = bitstream[i:]

        first_values = 2**b - self.m
        decimal = self.binary_to_decimal(binary_code)
        if decimal < first_values:
            return decimal + q * self.m if not negative else -1 * (decimal + q * self.m)
        else:
            return decimal + self.m - 2**b + q * self.m if not negative else -1 * (decimal + self.m - 2**b + q * self.m)

    def quocient(self, n, m):
        return math.floor(n / m)

    def remainder(self, n, m):
        return n % m
    
    def unary_code(self, q):
        return [1 for i in range(q)] + [0]
    
    def decimal_to_binary(self, decimal, bits):
        binary = []
        n = decimal
        while True:
            q = math.floor(n / 2)
            bit = n % 2
            binary.append(bit)
            n = q
            if q == 0:
                break
        
        binary = binary[::-1]
        if len(binary) < bits:
            binary = [0 for i in range(bits - len(binary))] + binary        
        return binary
    
    def binary_to_decimal(self, binary):
        return sum([int(binary[i]) * 2**(len(binary) - 1 - i) for i in range(len(binary))])
    
    def histogram(self, text):
        frequency = dict() 

        for symbol in text:
            if symbol not in frequency:
                frequency[symbol] = 1
            else:
                 frequency[symbol] += 1

        frequency = sorted(frequency.items(), reverse=True, key=lambda kv: kv[1])
        self.histogram = [(f[0], f[1] / len(text)) for f in frequency]
        return self.histogram


if __name__ == '__main__':
    golomb = Golomb(4)

    print(golomb.stream_decoder([0,0,0,0,0,0,0,1,0,0,1,0,0,0,1,1,0,1,0,0,0,0,1,0,0,1,0,1,0,1,0,0,1,0,1,1,0,1,1,0,0,0,0,1,1,0,0,1,0,1,1,0,1,0,0,1,1,0,1,1,0,1,1,1,0,0,0,0,1,1,1,0,0,1,0,1,1,1,0,1,1,0,1,1,1,0,1,0]))

File: cpp/src/golomb/Golomb.cpp
#include <assert.h>
#include <iostream>
#include <cstring>
#include <cmath>
#include <vector>
#include <algorithm>

using namespace std;

/**
 * This class implements the Golomb codification to compress information.
 * It encodes an integer based on a given value of M. 
 * The codification is the result of the concatenation of an unary code with a binary code,
 * given by the quotient and remainder of the division between integer value and M, respectively.
*/
class Golomb
{
private:
    double m;
    bool isBase2;

public:
    /**
         * Default constructor. 
         */
    Golomb()
    {
        this->m = 4; // default value
        this->isBase2 = true;
    }

    /**
         * Constructor with a given value of 'M'. 
         */
    Golomb(double m)
    {
        assert(m > 0);

        this->m = m;
        this->isBase2 = log2(this->m) == (int)log2(this->m) ? true : false;
    }

    /**
         * This method updates the value of 'M'.
         * 
         * @param m: new M
         */
    void set_m(double m)
    {
        assert(m > 0);

        this->m = m;
        this->isBase2 = log2(this->m) == (int)log2(this->m) ? true : false;
    }

    /**
         * Get the value of 'M'.
         */
    int get_m() { return this->m; }

    vector<int> encode(int n)
    {
        return this->isBase2 ? this->base2encoder(n) : this->truncated_encoder(n);
    }

    /**
         * This method encodes a given integer 'n' with Golomb encoding, using a power of two value of 'M'.
         * 
         * @param n: integer to encode
         */
    vector<int> base2encoder(int n)
    {

        //check 'n' sign (0 - positive, 1 - negative)
        vector<int> golomb_code;
        golomb_code.push_back(0);
        if (n < 0)
        {
            n = abs(n);
            golomb_code.at(0) = 1;
        }

        // computes values of 'q' and 'r'
        int q = floor(n / this->m);
        int r = n % (int)this->m;

        // computes unary and binary codes of 'q' and 'r', respectively.
        vector<int> unary_code = this->unary_code(q);
        vector<int> binary_code = this->binary_code(r, 2);

        golomb_code.insert(golomb_code.end(), unary_code.begin(), unary_code.end());
        golomb_code.insert(golomb_code.end(), binary_code.begin(), binary_code.end());

        return golomb_code;
    }

    /**
         * This method encodes a given integer 'n' with truncated Golomb encoding.
         * 
         * @param n: integer to encode
         */
    vector<int> truncated_encoder(int n)
    {
        int b = ceil(log2(this->m));

        int q = floor(n / this->m);
        int r = n % (int)this->m;

        vector<int> unary_code = this->unary_code(q);

        int first_values = pow(2, b) - this->m;

        vector<int> binary_code = r < first_values ? this->binary_code(r, b - 1) : this->binary_code(r + pow(2, b) - this->m, b);

        // golomb code
        unary_code.insert(unary_code.end(), binary_code.begin(), binary_code.end());

        return unary_code;
    }

    /*
        * This method, depending on the value of M, calls the respective method to decode a given list of bits.
        * If M is a power of two, it calls the base2decoder() method. Else, it calls the truncated_decoder() method.
        * 
        * @param bitstream: list of bits to decode
        */
    int decode(vector<int> bitstream)
    {
        assert(bitstream.size() > 0);
        return this->isBase2 ? this->base2decoder(bitstream) : this->truncated_decoder(bitstream);
    }

    /**
         * This method decodes a bitstream with multiple integers.
         * Example: stream_decoder([0,0,0,1,0,0,1,0,0,0,0,0]) -> [1,2,0]
         * 
         * @param bitstream: a list of bits 
        */

    vector<int> stream_decoder(vector<int> bitstream, int total = 0)
    {
        int bit = 0;

        // all decoded integers
        vector<int> decoded;

        if (bitstream.size() == 0)
            return decoded;

        // loop until all bits are decoded
        while (true)
        {
            try
            {
                int sign = bitstream.at(bit);
                bit++;
                vector<int> golomb_code = {sign};
                if (bitstream.at(bit) == 0)
                {
                    golomb_code.push_back(bitstream.at(bit)); // unary_code
                    bit++;
                    for (int i = bit; i < bit + ceil(sqrt(this->m)); i++)
                        golomb_code.push_back(bitstream.at(i));

                    bit += ceil(sqrt(this->m));
                    int decimal = this->decode(golomb_code);
                    decoded.push_back(decimal);
                }
                else
                {
                    while (true)
                    {
                        int unary_bit = bitstream.at(bit);
                        golomb_code.push_back(unary_bit);
                        bit++;
                        if (unary_bit == 0)
                            break;
                    }
                    for (int i = bit; i < bit + ceil(sqrt(this->m)); i++)
                        golomb_code.push_back(bitstream.at(i));

                    bit += ceil(sqrt(this->m));
                    int decimal = this->decode(golomb_code);
                    decoded.push_back(decimal);
                }

                if (bitstream.size() <= bit || decoded.size() == total)
                    break;
            }
            catch (...)
            {
                vector<int> eof = {0};
                return eof;
            }
        }
        return decoded;
    }

    /**
         * This method decodes a given list of bits, encoded with Golomb encoding.
         * 
         * @param bitstream: list of bits to decode
        */
    int base2decoder(vector<int> bitstream)
    {
        int bit = 0;
        bool negative = bitstream.at(bit) == 1;
        bit++;

        int q, r;
        if (bitstream.at(bit) == 0)
        {
            q = 0;
            vector<int> aux;
            for (int i = bit; i < bitstream.size(); i++)
                aux.push_back(bitstream.at(i));
            r = this->binary_to_decimal(aux);
        }
        else
        {
            int i = bit;
            while (bitstream.at(i) == 1)
                i++;
            i--;
            vector<int> binary_code;
            for (int j = i + 1; j < bitstream.size(); j++)
                binary_code.push_back(bitstream.at(j));
            q = i;
            r = this->binary_to_decimal(binary_code);
        }

        return negative ? -1 * (r + q * this->m) : r + q * this->m;
    }

    /**
         * This method decodes a given list of bits, encoded with truncated Golomb encoding.
         * 
         * @param bitstream: list of bits to decode
        */
    int truncated_decoder(vector<int> bitstream)
    {
        int b = ceil(log2(this->m));
        int q, r;
        vector<int> binary_code;
        if (bitstream.at(0) == 0)
        {
            q = 0;
            for (int i = 1; i < bitstream.size(); i++)
                binary_code.push_back(bitstream.at(i));
        }
        else
        {
            int i = 0;
            while (bitstream.at(i) == 1)
                i++;

            q = i;
            for (int j = i; j < bitstream.size(); j++)
                binary_code.push_back(bitstream.at(j));
        }
        int first_values = pow(2, b) - this->m;
        int decimal = this->binary_to_decimal(binary_code);
        return decimal < first_values ? decimal + q * this->m : decimal + this->m - pow(2, b) + q * this->m;
    }

    /**
         * This method, given a binary number, converts it into a natural number. 
        */
    int binary_to_decimal(vector<int> bitstream)
    {
        int decimal = 0;
        for (int i = 0; i < bitstream.size(); i++)
            decimal += bitstream.at(i) * pow(2, bitstream.size() - 1 - i);

        return decimal;
    }

    /**
         * This method, given an integer 'q', convert it into unary code of 1's followed by one 0. 
        */
    vector<int> unary_code(int q)
    {
        vector<int> unary_code;
        for (int i = 0; i < q; i++)
        {
            unary_code.push_back(1);
        }
        unary_code.push_back(0);
        return unary_code;
    }

    /**
         * This method, given a natural number, converts it into base2 with a given number of bits.
         * 
         * @param decimal: natural number
         * @param bits: number of bits
         */
    vector<int> binary_code(int decimal, int bits)
    {
        vector<int> binary;
        int n = decimal;

        while (true)
        {
            int q = floor(n / 2);
            int bit = n % 2;
            binary.push_back(bit);
            n = q;

            if (q == 0)
                break;
        }

        while (binary.size() < bits)
            binary.push_back(0);

        reverse(binary.begin(), binary.end());

        return binary;
    }
};
